spark/src/main/scala/org/apache/spark/rdd/FilteredCartesianRDD.scala

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+  * Adapted from [1] by Azavea, Inc.
+  *
+  * 1. https://github.com/apache/spark/blob/2f8776ccad532fbed17381ff97d302007918b8d8/core/src/main/scala/org/apache/spark/rdd/CartesianRDD.scala
+  */
+package org.apache.spark.rdd
+
+import java.io.{IOException, ObjectOutputStream}
+
+import scala.reflect.ClassTag
+
+import org.apache.spark._
+import org.apache.spark.util.Utils
+
+
+sealed class FilteredCartesianRDD[T: ClassTag, U: ClassTag, V: ClassTag](
+  sc: SparkContext,
+  pred: (T, U) => Boolean,
+  metapred: (V, V) => Boolean,
+  var rdd1 : RDD[T], metardd1 : RDD[V],
+  var rdd2 : RDD[U], metardd2 : RDD[V]
+) extends RDD[(T, U)](sc, Nil) with Serializable {
+
+  val numPartitionsInRdd2 = rdd2.partitions.length
+
+  override def getPartitions: Array[Partition] = {
+    // create the cross product split
+    val array = new Array[Partition](rdd1.partitions.length * rdd2.partitions.length)
+    for (s1 <- rdd1.partitions; s2 <- rdd2.partitions) {
+      val idx = s1.index * numPartitionsInRdd2 + s2.index
+      array(idx) = new CartesianPartition(idx, rdd1, rdd2, s1.index, s2.index)
+    }
+    array
+  }
+
+  override def getPreferredLocations(split: Partition): Seq[String] = {
+    val currSplit = split.asInstanceOf[CartesianPartition]
+    (rdd1.preferredLocations(currSplit.s1) ++ rdd2.preferredLocations(currSplit.s2)).distinct
+  }
+
+  override def compute(split: Partition, context: TaskContext): Iterator[(T, U)] = {
+    val currSplit = split.asInstanceOf[CartesianPartition]
+    val part1 = currSplit.s1 // partition for (|meta)rdd1
+    val part2 = currSplit.s2 // partition for (|meta)rdd2
+    val meta1 = metardd1.iterator(part1, context).next
```

Comment:
Just to document the conversation that I had with @echeipesh A little bit: when One does a `mapPartitions` with `preservesPartitioning = true`, the child RDD inherts its partitions directly from its parent (see [here](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/MapPartitionsRDD.scala#L35) and [here](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/RDD.scala#L245-L260)).  We assume that the partitions stay valid for the life of the RDD, because if they did not then the [`CartesianRDD` type](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/rdd/CartesianRDD.scala#L35-L36) would probably not work.